src/chainrules.jl

```diff
@@ -0,0 +1,30 @@
+import ChainRulesCore: ProjectTo, NoTangent
+
+"""
+    ProjectTo(::Fill) -> ProjectTo{Fill}
+    ProjectTo(::Ones) -> ProjectTo{NoTangent}
+
+Most FillArrays arrays store one number, and so their gradients under automatic
+differentiation represent the variation of this one number. 
+
+The exception is those like `Ones` and `Zeros` whose type fixes their value,
+which have no graidient.
+"""
+ProjectTo(x::Fill{<:Number}) = ProjectTo{Fill}(; element = ProjectTo(getindex_value(x)), axes = axes(x))
+
+ProjectTo(x::AbstractFill{Bool}) = ProjectTo{NoTangent}()  # Bool is always regarded as categorical
+
+ProjectTo(x::Zeros) = ProjectTo{NoTangent}()
+ProjectTo(x::Ones) = ProjectTo{NoTangent}()
+
+function (project::ProjectTo{Fill})(dx::AbstractArray)
+    for d in 1:max(ndims(dx), length(project.axes))
+        size(dx, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(axes_x, size(dx)))
+    end
+    Fill(mean(dx), project.axes)  # Note that mean(dx::Fill) is optimised
```

Comment:
Do we need another rule for the constructor to multiply the mean dx by the length of the vector? Think of `x -> sum(Fill(x, 3))`.

Reply:
Yes, the equivalent of https://github.com/FluxML/Zygote.jl/blob/e6a86745d66b5974eaafa8a8f28bcd4b100374df/src/lib/array.jl#L17

If the constructor is close to where the Fill is used, then perhaps it's a little wasteful to first project like this, and then un-create. But not so serious.

Reply:
yes